### ISSUE TITLE
Remove cgo on macos

### ISSUE THREAD

**Issue by hajimehoshi** [hajimehoshi/ebiten#1162]
https://groups.google.com/forum/m/#!topic/golang-nuts/Wt6P0BYndvs

**Comment by mortenson** [comment]
Would this enable cross compilation, i.e. compiling mac binaries from linux?

**Comment by hajimehoshi** [comment]
Yes, theoretically :-)

**Comment by hajimehoshi** [comment]
Another nice reference CL: https://go-review.googlesource.com/c/go/+/227037/

**Comment by hajimehoshi** [comment]
As M1 appears, might this be in higher demand than before?

**Comment by hajimehoshi** [comment]
Hmm, GLFW is the toughest blocker so far.

**Comment by hajimehoshi** [comment]
https://github.com/progrium/macdriver sounds reasonable!

**Comment by hajimehoshi** [comment]
I want to prioritize this, as it is now impossible to build amd64 version from M1 macOS.

**Comment by hajimehoshi** [comment]
Hm, we should be able to use environment variables `CC`, `CFLAGS` or something for other architectures.

**Comment by hajimehoshi** [comment]
https://developer.apple.com/documentation/objectivec/objective-c_runtime?language=objc

**Comment by TotallyGamerJet** [comment]
https://gist.github.com/vaiorabbit/6204931#file-building-libglfw-dylib-on-macos-apple-silicon

This explains how to build glfw as a dylib which can then by dynamically linked to using dlopen and dlsym.

**Comment by hajimehoshi** [comment]
Oops, I used a wrong issue number...

**Comment by aprchen** [comment]
> I want to prioritize this, as it is now impossible to build amd64 version from M1 macOS.

Is it can be compiled in version 2.4 ?

**Comment by hajimehoshi** [comment]
> Is it can be compiled in version 2.4 ?

Hopefully yes

**Comment by TotallyGamerJet** [comment]
In my Ebitengine branch [glfw-purego](https://github.com/TotallyGamerJet/ebiten/tree/glfw-purego) I Dlopen glfw as a Dylib. It is fully functional. I am trying to figure out the best way to distribute the dylib. When you did the dll for windows you copied it to the cache and opened it there. On macOS the traditional place to put dylib is in the .app inside some kind of Frameworks folder. Requiring users of Ebitengine to do that would not be backwards compatible bc some users use the direct executable and don’t use a .app. Perhaps write the file in a Frameworks folder no matter where the executable is located that way if it is in a .app it’ll be write and users can do externaldylib if they want to do it themselves and save space. Lmk what you think

**Comment by TotallyGamerJet** [comment]
Also when you did the dll you wrote a ten.go to get the dll. I used the script linked above to compile the dylib. Should I write a generator in Go as well?

**Comment by TotallyGamerJet** [comment]
According to [This post](https://developer.apple.com/forums/thread/125796) Apple will reject apps that don’t wrap .dylib in a .framework? Can we find an official source to validate this?

**Comment by hajimehoshi** [comment]
Hi, thank you for continuing this work!

Can we map the constent of the dysym and load it dynamically? (On Windows it was not possible and I had to extract a DLL as a file)

> Should I write a generator in Go as well?

Yes!

**Comment by hajimehoshi** [comment]
Another way is to translate GLFW to Go like what I did for Windows (https://github.com/hajimehoshi/ebiten/tree/main/internal/glfwwin)

**Comment by TotallyGamerJet** [comment]
Even with the parts that are the same that would be a lot to code especially since there are quite a few Objective-C classes that would all have to be ported. I was planning on working on that port after everything else with moved to purego but there doesn’t appear to be an easy way Dlopen/Dlsym a file from memory.

if I go the porting route should I just modify glfwwin package since there are constants and types that are shared?

**Comment by hajimehoshi** [comment]
> if I go the porting route should I just modify glfwwin package since there are constants and types that are shared?

Yes. I'm totally fine with renaming from `glfwwin` to, for example, `goglfw`. I keep the file structure as the original as much as possible.

**Comment by hajimehoshi** [comment]
Acknowledged. I don't think the overhead of SyscallN matters so much, and I'm fine with slowing a little. Thank you for your hard work!

**Comment by hajimehoshi** [comment]
As Go 1.20 will come soon, we would not make it for Ebitengine v2.5. I'll set the milestone to v2.6.

**Comment by hajimehoshi** [comment]
https://github.com/ebitengine/purego/issues/143 is the current blocker.

**Comment by hajimehoshi** [comment]
There can be a trouble around Xcode versions: 
https://wails.io/docs/guides/troubleshooting/#my-mac-app-gives-me-weird-compilation-errors

So, removing Cgo is very beneficial :D

**Comment by hajimehoshi** [comment]
As I'm pretty busy with other tasks, I have not been able to work on this recently 😢 . I think  https://github.com/ebitengine/purego/pull/175 could be a possible blocker, but is this actually? @TotallyGamerJet

**Comment by TotallyGamerJet** [comment]
It is a blocker for completion since there are a few C functions that return or take structs. The majority of it can be completed. Plus, struct Args PR is ready for review

**Comment by zakirullin** [comment]
Hi there! Thanks for such a good project.

What's the status of CGO removal for macOS?

**Comment by hajimehoshi** [comment]
No progress so far. Also, PureGo still lacks some features like https://github.com/ebitengine/purego/issues/225

**Comment by erexo** [comment]
cross compilation would be huge, I'd be able to build all versions of my game on my local machine

**Comment by Garguretke** [comment]
Bump — cross-compilation would be huge for game devs. Being able to build all platform targets locally without separate machines or VMs would simplify the release process enormously.

**Comment by hajimehoshi** [comment]
> Being able to build all platform targets locally without separate machines or VMs would simplify the release process enormously.

IIUC, VMs or CIs are still required for notarization, especially for macOS, even if Ebitengine becomes pure Go for macOS.

**Comment by TotallyGamerJet** [comment]
> IIUC, VMs or CIs are still required for notarization, especially for macOS, even if Ebitengine becomes pure Go for macOS.

There are programs that can do it in a cross platform way like [quill](https://github.com/anchore/quill)

**Comment by hajimehoshi** [comment]
> There are programs that can do it in a cross platform way like [quill](https://github.com/anchore/quill)

Oh, I would not try this, but that's good to know.

**Comment by hajimehoshi** [comment]
Done!

**Comment by zakirullin** [comment]
@hajimehoshi  Thank you a lot for your effort! I love the engine you've built for its radical simplicity and minimalism.

I build complex things on the engine, and I enjoy it quite a bit.

**Comment by TotallyGamerJet** [comment]
CMD+Q no longer quits the app

**Comment by hajimehoshi** [comment]
@TotallyGamerJet Thanks https://github.com/hajimehoshi/ebiten/issues/3413
